### PR TITLE
166575163 Borrow a book endpoint

### PR DESCRIPTION
**Pull request by allebd**
#### What does this PR do?
It allows users to borrow a book endpoint

#### Description of Task to be completed?
- create borrow a book validator
- create a route to borrow a book
- add authorization and authentication
- create a borrow book controller
- create test cases
- add documentation

#### How should this be manually tested?
- Use **git pull** to pull branch to local branch
- Use **npm install** to download all necessary packages
- Update default config file to use environment variables of the project
- Ensure your database is created or you can use **sequelize db:create** if database does not exist
- Ensure that there is **no id** defined in any models
- Use **npm run db:roll** to create table in database
- Use **npm run test**  to test
- Go your POSTMAN and type **http://localhost:${PORT}/api/v1/borrow**

#### Any background context you want to provide?
NA

#### What are the relevant pivotal tracker stories?
[#166575163](https://www.pivotaltracker.com/story/show/166575163)

#### Screenshots (if appropriate)
borrowedCapture
Sample of successful tests on borrowing a book route

borrownotfound
Sample of not successful test on borrowing a book route (book not found)

borrowalready

Sample of not successful test on borrowing a book route (book already borrowed)
#### Questions:
NA

### CODE REVIEW

File: server/controllers/UserController.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      if (!existingBook) return responseMessage(response, 404, { message: 'book not found' });
+      const { dataValues } = existingBook;
+      const bookStock = dataValues.stock;
+      const { fullname } = dataValues.Author;
+      delete dataValues.authorId;
+      delete dataValues.Author;
+      if (!bookStock) return responseMessage(response, 200, { status: 'success', message: 'book is out stock', data: { ...dataValues, authorName: fullname } });
+      await LendingHistory.create({ duration, bookId, userId: id });
+      await Book.update({ stock: bookStock - 1 }, { where: { id: bookId } });
+      return responseMessage(response, 201, { status: 'success', message: 'book successfully borrowed', data: { ...dataValues, authorName: fullname } });
```
Comment (henryade) [inline server/controllers/UserController.js]: what about if the stock is one and you subtract one and it turns to zero?

File: server/controllers/UserController.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      const { id } = request.userData;
+      const { duration, bookId } = request.body;
+      const existingBook = await Book.findOne({ where: { id: bookId }, include: [Author] });
+      if (!existingBook) return responseMessage(response, 404, { message: 'book not found' });
+      const { dataValues } = existingBook;
+      const bookStock = dataValues.stock;
+      const { fullname } = dataValues.Author;
+      delete dataValues.authorId;
+      delete dataValues.Author;
+      if (!bookStock) return responseMessage(response, 200, { status: 'success', message: 'book is out stock', data: { ...dataValues, authorName: fullname } });
```
Comment (henryade) [inline server/controllers/UserController.js]: this will never do anything.

File: server/controllers/UserController.js (hunk 1 above)
Comment (henryade) [inline server/controllers/UserController.js]: what about price?

File: server/middlewares/BorrowValidator.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+export default class BorrowValidator {
+  /**
+   * Generic Number validator
+   * @param {string} item
+   * @returns {function} call to a check API middleware
+   * @memberof Validation
+   */
+  static checkNumber(item) {
+    return BookValidator.checkNumber(item);
+  }
```
Comment (henryade) [inline server/middlewares/BorrowValidator.js]: you can just call the book validator down. you can remove this.

File: swagger.json [hunk 4]
```diff
@@ -137,6 +133,44 @@
...
+        ],
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [{
+          "in": "body",
+          "name": "body",
+          "description": "Borrow a book parameter",
+          "required": true,
+          "schema": {
```
Comment (henryade) [inline swagger.json]: no token?

**Review by henryade — CHANGES_REQUESTED** [review]

File: server/controllers/UserController.js (hunk 1 above)
Comment (allebd) [reply server/controllers/UserController.js]: When it turns to zero, it has been taken care of above, it would give a response book is out of stock

File: server/controllers/UserController.js (hunk 2 above)
Comment (allebd) [reply server/controllers/UserController.js]: Kindly test this route, I have tested it and it does something or why do you think it would not do anything

File: server/controllers/UserController.js (hunk 1 above)
Comment (allebd) [reply server/controllers/UserController.js]: The price is in the dataValues, it will also be displayed

File: server/middlewares/BorrowValidator.js (hunk 3 above)
Comment (allebd) [reply server/middlewares/BorrowValidator.js]: Thank you, noted

File: swagger.json (hunk 4 above)
Comment (allebd) [reply swagger.json]: Thank you, noted

File: server/controllers/UserController.js (hunk 1 above)
Comment (Adekoreday) [reply server/controllers/UserController.js]: when creating the lending history you didn't add a default status in the lending history table

**Review by Adekoreday — DISMISSED** [review]
add status to the lendinghistory table when you create a lending history

**Review by Ukhu — CHANGES_REQUESTED** [review]
Your branch is having conflicts with develop. You need to rebase.

File: test/server/routes/__mocks__/borrowData.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+export const newBookToBorrow = {
```
Comment (Ukhu) [inline test/server/routes/__mocks__/borrowData.js]: You should transfer the objects in this file to the bookData.js file

File: test/server/routes/__mocks__/index.js [hunk 6]
```diff
@@ -1,9 +1,11 @@
...
 import * as userData from './userData';
 import * as bookData from './bookData';
+import * as borrowData from './borrowData';
 
 export default {
   authorData,
   userData,
-  bookData
+  bookData,
+  borrowData
```
Comment (Ukhu) [inline test/server/routes/__mocks__/index.js]: No need for a borrowData file. Just put the objects in the bookData file

File: test/server/routes/__mocks__/borrowData.js (hunk 5 above)
Comment (allebd) [reply test/server/routes/__mocks__/borrowData.js]: Noted, done

File: test/server/routes/__mocks__/index.js (hunk 6 above)
Comment (allebd) [reply test/server/routes/__mocks__/index.js]: Noted, Done

File: server/helpers/findLendingHistory.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import jwt from 'jsonwebtoken';
+import models from '../database/models';
+import responseMessage from './responseMessage';
+
+const { LendingHistory } = models;
+
+const findLendingHistory = async (id, bookId, response) => {
+  try {
+    let bookDuration = '';
+    const lentBook = await LendingHistory.findOne({ where: { userId: id, bookId, type: 'borrowed' }, order: [['createdAt', 'DESC']], limit: 1 });
```
Comment (Lundii) [inline server/helpers/findLendingHistory.js]: I can see there is a column called type in the LendingHistory table. What other value does it hold apart from 'borrowed' and where was it used in this PR

Comment (allebd) [reply server/helpers/findLendingHistory.js]: Books can be borrowed or bought, but this route is for borrowed books only

**Review by Lundii — CHANGES_REQUESTED** [review]
Good job so far. Although there is a lot of file changed. I was able to successfully test it on my local. Can you add the duration of the borrowed book to the response object

Done

File: server/helpers/findLendingHistory.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    if (lentBook) {
+      const { durationToken, durationStatus } = lentBook.dataValues;
+      if (durationStatus !== 'expired') {
+        jwt.verify(durationToken, process.env.JWT_KEY, async (error) => {
+          if (error) {
+            bookDuration = 'book token expired';
+            await LendingHistory.update({ durationStatus: 'expired' }, { where: { userId: id, bookId } });
```
Comment (Lundii) [inline server/helpers/findLendingHistory.js]: Why are you updating the lending history. i thought your route is just to borrow a book. I think there should a route for the user to read a borrowed book. If the user tries to read a borrowed book, the route would check the duration token and if is expired, it should update the LendingHistory table. What do you think?

Comment (allebd) [reply server/helpers/findLendingHistory.js]: Thinking about it, you are also right, but for you to read a book, you have to borrow the book or check if the book has been borrowed, which I feel is this route

Comment (JuwonAbiola) [reply server/helpers/findLendingHistory.js]: yeah you're right

Comment (Lundii) [reply server/helpers/findLendingHistory.js]: Since Its a digital library, I don't think you need to know if the book is already borrowed because it is unlimited. Immediately the user successfully borrows the book, the session starts counting. Anytime the user logs in and requests to read a book, another route is called using fetch api which checks if the durationtoken is expired and updates the lending history table if it is.
This should probably be on another task though since this is focused on borrowing books

**Comment by Lundii** [comment]
> Good job so far. Although there is a lot of file changed. I was able to successfully test it on my local. Can you add the duration of the borrowed book to the response object

Look at this comment I made earlier

Noted, Done

File: server/helpers/findLendingHistory.js (hunk 7 above)
Comment (Nennyfills) [inline server/helpers/findLendingHistory.js]: The lentBook variable you named this is quite confusing, shouldn't it be lendBook, since you're getting a book from LendingHistory database, looks like you made changes on every single file.

**Review by Nennyfills — DISMISSED** [review]
The lentBook variable you named this is quite confusing, shouldn't it be lendBook, since your getting a book from LendingHistory database,

Noted, treated accordingly

File: server/helpers/findLendingHistory.js (hunk 8 above)
Comment (allebd) [reply server/helpers/findLendingHistory.js]: I was earlier told to add this implementation by the TTL to update and check for expired lent books

Comment (Lundii) [reply server/helpers/findLendingHistory.js]: Okay. I will bring it up during the standup. Nice work

File: server/helpers/findLendingHistory.js (hunk 7 above)
Comment (allebd) [reply server/helpers/findLendingHistory.js]: The reason I used the word lent is that the book has been lent, would rename it to borrowedBook for better clarification.

Regarding the file changes.
- Borrowing a book affected lendingHistory, users and all their respective locations.
- Fines and Followers were removed because they were not needed again and it affected all their respective locations.
- Favourites had two different spellings in the code base and that was renamed for consistency and affected all their respective locations.
- Author fullname which I earlier created was renamed to authorName because of the fullname was ambiguous which affected all its locations.
- These are the reasons for the many file changes.

File: server/helpers/findLendingHistory.js (hunk 8 above)
Comment (allebd) [reply server/helpers/findLendingHistory.js]: Updated accordingly, thank you for the observation

**Review by Lundii — DISMISSED** [review]

File: server/controllers/BookController.js [hunk 9]
```diff
@@ -22,7 +22,8 @@ class BookController {
       }
     } = request;
     try {
-      const newAuthor = await Author.findOrCreate({ where: { fullname: authorName } });
+      const fullname = makeLowerCase(authorName);
+      const newAuthor = await Author.findOrCreate({ where: { authorName: fullname } });
```
Comment (henryade) [inline server/controllers/BookController.js:26]: I believe its authorName you get from postman also

File: server/controllers/UserController.js [hunk 10]
```diff
@@ -155,10 +142,45 @@ export default class UserController {
...
+      if (borrowedBook && bookDuration === '') responseMessage(response, 409, { message: 'book already borrowed' });
+      const borrowingBook = await LendingHistory.create({
+        bookId, userId: id, charge: SERVICE_CHARGE
+      });
+      const { charge, duration } = borrowingBook;
+      responseMessage(response, 201, {
+        status: 'success',
+        message: 'book successfully borrowed',
+        data: {
+          ...dataValues, bookAuthor: authorName, charge, duration
```
Comment (henryade) [inline server/controllers/UserController.js]: should be author not  bookAuthor

File: server/database/seeders/20190614115108-demo-lending-histories.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable */
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert('LendingHistories', [{
+    userId: 3,
+    bookId: 3,
+    charge: 200,
+    durationToken: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MywiaWF0IjoxNTYwNzQxNTMyfQ.rLSG383b3t9ywzpYP1di-9hAePJxJglF2Nz0O-4L9oU',
```
Comment (henryade) [inline server/database/seeders/20190614115108-demo-lending-histories.js]: your jwt key might be different from another person own. the person will not be able to decode it to test the end point

File: server/helpers/findLendingHistory.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+const findLendingHistory = async (id, bookId, response) => {
+  try {
+    let bookDuration = '';
+    const borrowedBook = await LendingHistory.findOne({ where: { userId: id, bookId, type: 'borrowed' }, order: [['createdAt', 'DESC']], limit: 1 });
+    if (borrowedBook) {
+      const { durationToken } = borrowedBook.dataValues;
+      jwt.verify(durationToken, process.env.JWT_KEY, async (error) => {
+        if (error) {
+          bookDuration = 'book token expired';
```
Comment (henryade) [inline server/helpers/findLendingHistory.js:15]: what about if the token isnt expired but it cant decode the token?

File: server/database/migrations/20190602223101-create-lending-history.js [hunk 13]
```diff
@@ -24,13 +19,32 @@ module.exports = {
...
+      },
+      type: {
+        allowNull: false,
+        type: Sequelize.STRING,
+        defaultValue: 'borrowed'
+      },
+      duration: {
+        allowNull: false,
+        type: Sequelize.STRING,
+        defaultValue: '1 month'
```
Comment (henryade) [inline server/database/migrations/20190602223101-create-lending-history.js:42]: does this value change?

File: server/middlewares/UserValidator.js [hunk 14]
```diff
@@ -231,7 +245,7 @@ export default class UserValidator {
       UserValidator.checkEmail(),
       UserValidator.checkName('firstName'),
       UserValidator.checkName('lastName'),
-      UserValidator.checkRole(),
+      UserValidator.checkRole('roleId'),
```
Comment (henryade) [inline server/middlewares/UserValidator.js]: the roleId parameter is not needed

**Review by henryade — CHANGES_REQUESTED** [review]
What happens if a books borrowing period expires? or will that be another task?

When a borrowed book expires, the durationToken stored would automatically be expired

**Review by Ukhu — COMMENTED** [review]
The screenshot for borrow book isn't showing the status code

okay, would work on that

File: .env.sample [hunk 15]
```diff
@@ -11,7 +11,8 @@ JWT_KEY=
 SENDGRID_API_KEY=
 ADMIN_MAIL=
 PASSWORD=
-BASE_URL=
```
Comment (Ukhu) [inline .env.sample:14]: why did you remove this?

File: server/controllers/BookController.js (hunk 9 above)
Comment (allebd) [reply server/controllers/BookController.js:26]: yes it is authorName, but I want to convert the request to lower case that is why it was later changed to fullname

File: server/controllers/UserController.js (hunk 10 above)
Comment (allebd) [reply server/controllers/UserController.js]: Based on the TTL feedback that is why it was named that but the author also sounds good too

File: server/database/seeders/20190614115108-demo-lending-histories.js (hunk 11 above)
Comment (allebd) [reply server/database/seeders/20190614115108-demo-lending-histories.js]: Actually, this does not affect this PR test in any way, it would be removed

File: server/helpers/findLendingHistory.js (hunk 12 above)
Comment (allebd) [reply server/helpers/findLendingHistory.js:15]: If it can't decode the token, it will be treated as a new request and a new token would be generated.

File: server/database/migrations/20190602223101-create-lending-history.js (hunk 13 above)
Comment (allebd) [reply server/database/migrations/20190602223101-create-lending-history.js:42]: Yes, in case a patron chooses to extend the duration

File: server/middlewares/UserValidator.js (hunk 14 above)
Comment (allebd) [reply server/middlewares/UserValidator.js]: True, would work on that

File: .env.sample (hunk 15 above)
Comment (allebd) [reply .env.sample:14]: BASE_URL is repeated throughout the project for different routes, using it as the same name for the default SERVER URL, when being called on a similar page, it might cause a conflict

File: server/controllers/BookController.js (hunk 9 above)
Comment (henryade) [reply server/controllers/BookController.js:26]: but when you are adding a book, i beleive authorname is more descriptive than fullname

Comment (allebd) [reply server/controllers/BookController.js]: corrected

Comment (allebd) [reply server/controllers/BookController.js:26]: yes it is

**Review by Lundii — DISMISSED** [review]

**Review by Ukhu — COMMENTED** [review]
Rebase branch so I can approve

File: server/controllers/UserController.js [hunk 16]
```diff
@@ -155,10 +142,45 @@ export default class UserController {
...
+    try {
+      const { id, paymentStatus } = request.userData;
+      const { bookId } = request.body;
+      if (!paymentStatus) responseMessage(response, 402, { message: 'insufficient funds to borrow book' });
+      const existingBook = await findBook(bookId, response);
+      const { dataValues, BookAuthors: [BookAuthor] } = existingBook;
+      const { authorName } = BookAuthor.dataValues.Author;
+      delete dataValues.BookAuthors;
+      const [borrowedBook, bookDuration] = await findLendingHistory(id, bookId, response);
+      if (borrowedBook && bookDuration === '') responseMessage(response, 409, { message: 'book already borrowed' });
```
Comment (Ukhu) [inline server/controllers/UserController.js]: You need to use the 'return' keyword on this line otherwise it will lead to this error: _Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client_

Comment (allebd) [reply server/controllers/UserController.js]: Noted

**Review by Lundii — APPROVED** [review]

**Review by Ukhu — APPROVED** [review]

**Review by henryade — APPROVED** [review]

**Review by JuwonAbiola — APPROVED** [review]
